step2 구현

shinyeonjong opened the pull request.

shinyeonjong committed:
step2 구현

phs1116 commented on src/main/java/nextstep/service/QnaService.java:
```diff
@@ -31,19 +31,22 @@ public Question create(User loginUser, Question question) {
-        return null;
+    public Question update(User loginUser, long questionId, Question updatedQuestion) {
+        Question question = findById(questionId);
+        question.writeBy(question.getWriter());
```
question writer를 다시 셋팅할 할 필요가 있을까요?

phs1116 commented on src/main/java/nextstep/web/HomeController.java:
```diff
@@ -1,13 +1,21 @@
 @Controller
 public class HomeController {
+    @Autowired
+    QnaService qnaService;
```
캡슐화를 지켜주시면 좋을 것 같습니다. :)

phs1116 commented on src/main/java/nextstep/web/QuestionController.java:
```diff
@@ -0,0 +1,62 @@
+    private QnaService qnaService;
+
+    @GetMapping("/form")
+    public String form() {
```
`로그인한 사용자에 대해서만 질문이 가능하다.`
로그인을 하지 못한 사람도 작성 폼을 볼 수 있을 것 같습니다. :)

phs1116 commented on src/test/java/nextstep/web/QuestionAcceptanceTest.java:
```diff
@@ -0,0 +1,113 @@
+    public void delete_success() {
+        Question question = questionRepository.findById(UserTest.JAVAJIGI.getId()).get();
+        ResponseEntity<String> response = this.delete(basicAuthTemplate(), question);
+        softly.assertThat(response.getStatusCode()).isEqualTo(HttpStatus.UNAUTHORIZED);
```
delete_fail, delete_success가 둘 다 실패에 대한 테스트인 것 같네요. :)

phs1116 commented on src/test/java/nextstep/web/QuestionAcceptanceTest.java:
```diff
@@ -0,0 +1,113 @@
+    private QuestionRepository questionRepository;
+
+    @Test
+    public void form() {
```
실패 케이스에 대한 테스트도 작성하면 어떨까요?

phs1116 commented on src/test/java/nextstep/domain/QuestionTest.java:
```diff
@@ -0,0 +1,55 @@
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+public class QuestionTest extends BaseTest {
```
엔티티 외에 QnaService등 Service 레이어도 테스트를 추가해보면 어떨까요?

phs1116 commented on src/test/java/nextstep/web/QuestionAcceptanceTest.java:
```diff
@@ -0,0 +1,113 @@
+
+    private ResponseEntity<String> delete(TestRestTemplate template, Question question) {
+        HttpEntity<MultiValueMap<String, Object>> request = HtmlFormDataBuilder.urlEncodedForm()
+                .addParameter("_method", "delete")
```
addParameter("_method", "put"), addParameter("_method", "delete")도 좋지만 하드코딩 하지 말고 HtmlFormDataBuilder에 해당 기능을 구현해보면 어떨까요?

phs1116 reviewed (changes requested):
안녕하세요. 리뷰를 맡게된 리뷰어 박현식입니다. :) 
전체적으로 구현 잘하셨어요. 👍 
피드백 몇가지 추가했습니다. 참고 부탁드립니다.
또한 궁금하신 점 있으시면 연락주세요!!

shinyeonjong committed:
step2 피드백 반영
1. HomeController
- - qnaService 캡슐화
2. QuestionController
- 로그인한 사용자에 대해서만 접근하도록 수정
3. QuestionAcceptanceTest
- 삭제 테스트 오류 수정
- form 실패 테스트 작성
4. QnaService Test 추가
5. HtmlFormDataBuilder
- put, delete 메소드 추가

phs1116 reviewed (approved):
안녕하세요. 리뷰어 박현식입니다. :)
피드백 반영해 주셔서 감사합니다.
2단계 머지할게요. :) 개발하시느라 수고 많으셨습니다!
